include/rnp/rnp2.h

```diff
@@ -0,0 +1,241 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*
+* Currently we use some definitions/enums from other others, including
+* - rnp_result
+* - pgp_hash_alg_t
+* - pgp_pubkey_alg_t
+*
+* It might be good to consolidate these definitions into one spot, or
+* else replicate them here in new enums and make this header freestanding.
+*/
+#include <rnp_def.h>
+#include <repgp_def.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+/*
+* Opaque structures for key types
+*/
+typedef struct rnp_publickey_st * rnp_public_key;
+typedef struct rnp_privatekey_st *rnp_private_key;
+
+// typedef struct rnp_keyring_st* rnp_keyring;
+
+/* Operations on public keys */
+
+/**
+* Load a public key
+* @param key a pointer, on success will be assigned to the new object
+* @param buf the serialized key (either binary or ASCII formats accepted)
+* @param buf_len length of buf in bytes
+*/
+rnp_result rnp_public_key_load(rnp_public_key *key, const uint8_t buf[], size_t buf_len);
+
+/**
+* @param key the public key
+* @param uid on success will be set to the UID of the key
+*/
+rnp_result rnp_public_key_get_uid(rnp_public_key key, char **uid);
+
+/**
+* @param key the public key
+* @param fprint on success will be set to the fingerprint of the key
+*/
+rnp_result rnp_public_key_get_fprint(rnp_public_key key, char **fprint);
+
+/* Operations on private keys */
+
+/**
+* Callback used for getting a passphrase.
+* @param app_ctx provided by application
+* @param key the key that is being decrypted
+* @param pass to which the callback should write the returned
+* passphrase, NULL terminated.
+* @param pass_len the size of pass buffer
+* @return 0 on success, or any other value to stop decryption.
+*/
+typedef int (*rnp_passphrase_cb)(void *          app_ctx,
+                                 rnp_private_key key,
+                                 char            buf[],
+                                 size_t          buf_len);
+
+/*
+* Load a private key into memory. Initially it will be encrypted,
+* assuming the input is encrypted.
+*/
+rnp_result rnp_private_key_load(rnp_private_key *key, const uint8_t buf[], size_t buf_len);
```

Comment:
I wonder how we should handle different formats here. Typically it would be PGP or G10. We could just auto-detect internally (G10 starts and ends with parentheses, so it's not like it's difficult to guess).

Reply:
Or we could extract out an additional function called `rnp_detect_key_format`? Too much magic could make like difficult.